jax/interpreters/pxla.py

```diff
@@ -899,27 +946,37 @@ def _pval_to_result_handler(axis_size, nrep, pval, devices, backend):
   else:
     if pv is not core.abstract_unit:
       unsharded_aval = ShapedArray((axis_size,) + pv.shape, pv.dtype)
-      sharding_spec = _pmap_sharding_spec(nrep, axis_size, pv, True)
+      sharding_spec = _pmap_sharding_spec(nrep, axis_size, npart, parts, pv,
+                                          True)
       indices = spec_to_indices(unsharded_aval.shape, sharding_spec)
     else:
       sharding_spec = indices = None
       unsharded_aval = pv
     return aval_to_result_handler(sharding_spec, indices, unsharded_aval)
 
-def _pmap_sharding_spec(nrep, axis_size, sharded_aval, mapped):
+def _pmap_sharding_spec(nrep, axis_size, npart, parts, sharded_aval, mapped):
+  if not mapped and npart > 1:
+    # TODO(skye): ShardingSpec assumes replication is treated as the innermost
+    # axis, but in this case, it needs to be the outer axis.
```

Comment:
Maybe what we want is two replication factors, one that will map to XLA replicas and the other to XLA partitions? I'll look into that option (and in general I can take this TODO; if I get WUS working and you get pods working, I think we'll be in business)

Reply:
Added your name to the TODO :) I think two replication factors would work for this case, but not the most general case of in_axes/out_axes=N for N > 1 on pmap. Introducing a new replication factor until we figure out a more general solution seems strictly better than the current situation though, and doesn't seem like too much work (I still have staging around `ShardedDeviceArray.__init__` so we don't have to update all callers when we update how sharding works).